File: protocol/src/main/resources/csharp/ProtocolTemplate.cs
using System;
using System.Collections.Generic;

namespace zfoocs
{
    {}
    public class {}
    {
        {}
    }


    public class {}Registration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return {};
        }

        public void Write(ByteBuffer buffer, object packet)
        {
            if (packet == null)
            {
                buffer.WriteInt(0);
                return;
            }
            {} message = ({}) packet;
            {}
        }

        public object Read(ByteBuffer buffer)
        {
            int length = buffer.ReadInt();
            if (length == 0)
            {
                return null;
            }
            int beforeReadIndex = buffer.ReadOffset();
            {} packet = new {}();
            {}
            if (length > 0) {
                buffer.SetReadOffset(beforeReadIndex + length);
            }
            return packet;
        }
    }
}

File: protocol/src/main/resources/python/ProtocolTemplate.py
${protocol_note}
class ${protocol_name}:
    ${protocol_field_definition}

    def protocolId(self):
        return ${protocol_id}

    @classmethod
    def write(cls, buffer, packet):
        if packet is None:
            buffer.writeInt(0)
            return
        ${protocol_write_serialization}
        pass

    @classmethod
    def read(cls, buffer):
        length = buffer.readInt()
        if length == 0:
            return None
        beforeReadIndex = buffer.getReadOffset()
        packet = ${protocol_name}()
        ${protocol_read_deserialization}
        if length > 0:
            buffer.setReadOffset(beforeReadIndex + length)
        return packet